accepted/future-releases/records/records-feature-specification.md

```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
 
 The greatest lower bound of records with different shapes is `Never`.
 
-### Type inference and promotion
+### Type inference
+
+As usual, we define type inference for record expressions with respect to a
+context type schema which is determined by the surrounding context of the
+inferred expression.  Given a type schema `K` and a record expression `E` of the
+general form `(e0, ..., en, d0 : t0, ..., dm : tm)`, then inference proceeds as
+follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : P0, ...., dm :
+Pm})`, then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+  - Each `ti` is inferred with context type schema `Pi` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+If `K` is any other type schema:
+  - Each `ei` is inferred with context type schema `_` to have type `Ti`
+  - Each `ti` is inferred with context type schema `_` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+Note that contrary to the practice for runtime checked covariant nominal types,
+we do not prefer the context type over the more precise upwards type.  The
+following example illustrates this:
+
+```dart
+  // No error, inferred type of the record is (int, double)
+  (num, num) r = (3, 3.5)..$0.isEven;
+```
 
-Type inference and promotion flows through records in much the same way it does
-for instances of generic classes (which are covariant in Dart just like record
-fields are) and collection literals.
+Note that implicit casts are not considered to be applied as part of inference,
```

Comment:
That's news to me. (Although I'll admit I wouldn't be able to point to where they are specified at all. I just always thought that the implicit downcasts were inserted during inference, like the missing type parameters and missing `.call` member accesses.)

I guess it's an acceptable distinction, since we allow *assignability* from one to the other, and we presumably remember the static type and context type found during inference for the rest of the life of the program (the runtime-semantics sometimes refers to them).

Which means that the implicit downcast is part of *runtime* semantics, and does not affect static typing.

I'm guessing `call` member tear-off is the same,
but double-literal conversion is not, and neither is implicit instantiated tear-off (mainly for historical reasons, today it probably could be):
```dart
T id<T>(T value) => value;
int Function(int) f = id..printStaticType(); 
```
prints `(int) -> int`.